Subject expansion failed when depth exceeds 2

### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/keto/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/keto/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Network](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

Given the namespace definition
```ts
class Role implements Namespace {
	related: {
		perms: SubjectSet<Role, "perms">[]
	}
}
```
and the relation tuples:
```
"Role:a#perms@Role:b#perms",
"Role:b#perms@Role:c#perms",
"Role:c#perms@Role:d#perms",
"Role:d#perms@get",
```

The check "Role:a#perms@get" will fail.

### Reproducing the bug

```go
package check_test

import (
	"context"
	"testing"

	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"go.uber.org/goleak"

	"github.com/ory/keto/internal/check"
	"github.com/ory/keto/internal/check/checkgroup"
	"github.com/ory/keto/internal/driver/config"
	"github.com/ory/keto/internal/namespace"
	"github.com/ory/keto/internal/schema"
)

func TestSubjectExpansion(t *testing.T) {
	var rawns = `
	import { Namespace, SubjectSet, Context } from '@ory/keto-namespace-types';
	
	class Role implements Namespace {
	  related: {
		perms: SubjectSet<Role, "perms">[]
	  }
	}`

	nss, err := schema.Parse(rawns)
	require.True(t, err == nil)
	namespaces := make([]*namespace.Namespace, len(nss))
	for i, ns := range nss {
		ns := ns
		namespaces[i] = &ns
	}

	reg := newDepsProvider(t, namespaces)
	reg.Logger().Logger.SetLevel(logrus.InfoLevel)

	insertFixtures(t, reg.RelationTupleManager(), []string{
		"Role:a#perms@Role:b#perms",
		"Role:b#perms@Role:c#perms",
		"Role:c#perms@Role:d#perms",
		"Role:d#perms@get",
	})

	testCases := []struct {
		query         string
		expected      checkgroup.Result
		expectedPaths []path
	}{{
		query:    "Role:d#perms@get",
		expected: checkgroup.ResultIsMember,
	}, {
		query:    "Role:c#perms@get",
		expected: checkgroup.ResultIsMember,
	}, {
		query:    "Role:b#perms@get",
		expected: checkgroup.ResultIsMember,
	}, { // This case will fail
		query:    "Role:a#perms@get",
		expected: checkgroup.ResultIsMember,
	}}

	t.Run("suite=testcases", func(t *testing.T) {
		ctx := context.Background()
		reg.Config(ctx).Set(config.KeyLimitMaxReadDepth, 10)
		e := check.NewEngine(reg)
		defer goleak.VerifyNone(t, goleak.IgnoreCurrent())

		for _, tc := range testCases {
			t.Run("case="+tc.query, func(t *testing.T) {
				rt := tupleFromString(t, tc.query)

				res := e.CheckRelationTuple(ctx, rt, 100)
				require.NoError(t, res.Err)
				t.Logf("tree:\n%s", res.Tree)
				assert.Equal(t, tc.expected.Membership, res.Membership)

				if len(tc.expectedPaths) > 0 {
					for _, path := range tc.expectedPaths {
						assertPath(t, path, res.Tree)
					}
				}
			})
		}
	})
}
```

### Relevant log output

```shell
=== RUN   TestSubjectExpansion
time=2023-02-24T19:45:04+08:00 level=info msg=No tracer configured - skipping tracing setup audience=application service_name=Ory Keto service_version=testing
=== RUN   TestSubjectExpansion/suite=testcases
=== RUN   TestSubjectExpansion/suite=testcases/case=Role:d#perms@get
    /keto/internal/check/check_test.go:80: tree:
        ∋ Role:5d0ab373-34e6-5449-9217-196aa2ef5d76#perms@01338e40-ba98-5539-a9ee-856b0223c9b1️
=== RUN   TestSubjectExpansion/suite=testcases/case=Role:c#perms@get
    /keto/internal/check/check_test.go:80: tree:
=== RUN   TestSubjectExpansion/suite=testcases/case=Role:b#perms@get
    /keto/internal/check/check_test.go:80: tree:
=== RUN   TestSubjectExpansion/suite=testcases/case=Role:a#perms@get
    /keto/internal/check/check_test.go:80: tree:
    /keto/internal/check/check_test.go:81:
        	Error Trace:	/keto/internal/check/check_test.go:81
        	Error:      	Not equal:
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestSubjectExpansion/suite=testcases/case=Role:a#perms@get
--- FAIL: TestSubjectExpansion (0.02s)
    --- FAIL: TestSubjectExpansion/suite=testcases (0.00s)
        --- PASS: TestSubjectExpansion/suite=testcases/case=Role:d#perms@get (0.00s)
        --- PASS: TestSubjectExpansion/suite=testcases/case=Role:c#perms@get (0.00s)
        --- PASS: TestSubjectExpansion/suite=testcases/case=Role:b#perms@get (0.00s)
        --- FAIL: TestSubjectExpansion/suite=testcases/case=Role:a#perms@get (0.00s)
FAIL
FAIL	github.com/ory/keto/internal/check	0.811s
```


### Relevant configuration

_No response_

### Version

master

### On which operating system are you observing this issue?

None

### In which environment are you deploying?

None

### Additional Context

_No response_

--- a/.schema/version.schema.json
+++ b/.schema/version.schema.json
@@ -1,193 +1,171 @@
 {
-    "$id": "https://github.com/ory/keto/.schema/versions.config.schema.json",
-    "$schema": "http://json-schema.org/draft-07/schema#",
-    "oneOf": [
+  "$id": "https://github.com/ory/keto/.schema/versions.config.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "oneOf": [
+    {
+      "allOf": [
         {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.11.0-alpha.0"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.11.0-alpha.0/.schema/config.schema.json"
-                }
-            ]
+          "properties": {
+            "version": {
+              "const": "v0.11.0-alpha.0"
+            }
+          },
+          "required": ["version"]
         },
         {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.10.0-alpha.0"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.10.0-alpha.0/.schema/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.9.0-alpha.0"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.9.0-alpha.0/.schema/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.8.0-alpha.2"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.8.0-alpha.2/.schema/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.6.0-alpha.1"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.6.0-alpha.1/internal/driver/config/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.6.0-alpha.3"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.6.0-alpha.3/internal/driver/config/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.7.0-alpha.0"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.7.0-alpha.0/.schema/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.7.0-alpha.1"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.7.0-alpha.1/.schema/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "properties": {
-                        "version": {
-                            "const": "v0.8.0-alpha.1"
-                        }
-                    },
-                    "required": [
-                        "version"
-                    ]
-                },
-                {
-                    "$ref": "https://raw.githubusercontent.com/ory/keto/v0.8.0-alpha.1/.schema/config.schema.json"
-                }
-            ]
-        },
-        {
-            "allOf": [
-                {
-                    "oneOf": [
-                        {
-                            "properties": {
-                                "version": {
-                                    "type": "string",
-                                    "maxLength": 0
-                                }
-                            },
-                            "required": [
-                                "version"
-                            ]
-                        },
-                        {
-                            "not": {
-                                "properties": {
-                                    "version": {}
-                                },
-                                "required": [
-                                    "version"
-                                ]
-                            }
-                        }
-                    ]
-                },
-                {
-                    "$ref": "#/oneOf/0/allOf/1"
-                }
-            ]
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.11.0-alpha.0/.schema/config.schema.json"
         }
-    ],
-    "title": "All Versions of the ORY Keto Configuration",
-    "type": "object"
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.10.0-alpha.0"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.10.0-alpha.0/.schema/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.9.0-alpha.0"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.9.0-alpha.0/.schema/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.8.0-alpha.2"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.8.0-alpha.2/.schema/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.6.0-alpha.1"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.6.0-alpha.1/internal/driver/config/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.6.0-alpha.3"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.6.0-alpha.3/internal/driver/config/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.7.0-alpha.0"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.7.0-alpha.0/.schema/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.7.0-alpha.1"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.7.0-alpha.1/.schema/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "v0.8.0-alpha.1"
+            }
+          },
+          "required": ["version"]
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/ory/keto/v0.8.0-alpha.1/.schema/config.schema.json"
+        }
+      ]
+    },
+    {
+      "allOf": [
+        {
+          "oneOf": [
+            {
+              "properties": {
+                "version": {
+                  "type": "string",
+                  "maxLength": 0
+                }
+              },
+              "required": ["version"]
+            },
+            {
+              "not": {
+                "properties": {
+                  "version": {}
+                },
+                "required": ["version"]
+              }
+            }
+          ]
+        },
+        {
+          "$ref": "#/oneOf/0/allOf/1"
+        }
+      ]
+    }
+  ],
+  "title": "All Versions of the ORY Keto Configuration",
+  "type": "object"
 }

--- a/internal/check/engine.go
+++ b/internal/check/engine.go
@@ -137,7 +137,12 @@ func (e *Engine) checkExpandSubject(r *relationTuple, restDepth int) checkgroup.
 
 		// If not, we must go another hop:
 		for _, result := range results {
-			innerCtx, visited = graph.CheckAndAddVisited(innerCtx, result.To.Subject)
+			sub := &relationtuple.SubjectSet{
+				Namespace: result.To.Namespace,
+				Object:    result.To.Object,
+				Relation:  result.To.Relation,
+			}
+			innerCtx, visited = graph.CheckAndAddVisited(innerCtx, sub)
 			if visited {
 				continue
 			}

--- a/internal/check/engine_test.go
+++ b/internal/check/engine_test.go
@@ -15,6 +15,7 @@ import (
 	"github.com/ory/keto/internal/driver"
 	"github.com/ory/keto/internal/driver/config"
 	"github.com/ory/keto/internal/namespace"
+	"github.com/ory/keto/internal/namespace/ast"
 	"github.com/ory/keto/internal/persistence"
 	"github.com/ory/keto/internal/relationtuple"
 	"github.com/ory/keto/internal/x"
@@ -219,6 +220,50 @@ func TestEngine(t *testing.T) {
 		}, 0)
 		require.NoError(t, err)
 		assert.False(t, res)
+	})
+
+	t.Run("subject expansion", func(t *testing.T) {
+		reg := newDepsProvider(t, []*namespace.Namespace{{
+			Name: "n",
+			Relations: []ast.Relation{
+				{
+					Name: "r",
+					Types: []ast.RelationType{
+						{
+							Namespace: "n",
+							Relation:  "r",
+						},
+					},
+				},
+			},
+		}})
+		tuples := []string{
+			`n:a#r@n:b#r`,
+			`n:b#r@n:c#r`,
+			`n:c#r@n:d#r`,
+			`n:d#r@u`,
+		}
+
+		insertFixtures(t, reg.RelationTupleManager(), tuples)
+		e := check.NewEngine(reg)
+		reg.Config(ctx).Set(config.KeyLimitMaxReadDepth, 5)
+
+		cases := []struct {
+			tuple string
+		}{
+			{tuple: "n:d#r@u"},
+			{tuple: "n:c#r@u"},
+			{tuple: "n:b#r@u"},
+			{tuple: "n:a#r@u"},
+		}
+
+		for _, tc := range cases {
+			t.Run("case="+tc.tuple, func(t *testing.T) {
+				res, err := e.CheckIsMember(ctx, tupleFromString(t, tc.tuple), 0)
+				require.NoError(t, err)
+				assert.True(t, res)
+			})
+		}
 	})
 
 	t.Run("wrong object ID", func(t *testing.T) {

--- a/package.json
+++ b/package.json
@@ -20,7 +20,7 @@
     "@ory/keto-namespace-types": "file:contrib/namespace-type-lib",
     "doctoc": "^2.0.1",
     "license-checker": "^25.0.1",
-    "ory-prettier-styles": "^1.2.0",
+    "ory-prettier-styles": "1.3.0",
     "prettier": "2.8.4",
     "prettier-plugin-packagejson": "^2.2.18"
   }

